### ISSUE TITLE
Residual Google Analytics wildcard-domain cookies cause false positives

### ISSUE THREAD

**Issue by maranomynet** [EFForg/privacybadger#574]
Say I have two servers, `assets.example.com`and `www.example.com`.

`assets.example.com` only serves harmless (non-tracking) assets (JavaScripts, CSS, images) that several different sites link to and rely on. This server sets no cookies.

`www.example.com` is a simple homepage that has Google Analytics, and before being (rightfully) blocked by PrivacyBadger, Google Analytics set long-lived cookies to the wildcard-domain `.example.com`.

These cookies persist after the block and get sent along whenever the browser visits some of the sites that link to `assets.example.com`. This PrivacyBadger picks up on and promptly blocks `example.com` as an evil tracker.

In effect, PrivacyBadger's approach allows GoogleAnalytics to "infect" other sites.

---

What do you suggest - short of bying a new cookieless domain `asssets-example.com` and manually update potentially thousands of links on all the various 3rd party sites?

(HTTP 301 redirects won't do much good as they rely on `assets.example.com` not being blocked.)

**Comment by cooperq** [comment]
weird, assets.example.com should not be blocked unless it is loaded as a third party on three or more other domains - and certainly never on www.example.com .  This might point to a larger issue.

**Comment by maranomynet** [comment]
Clarification:  `assets.example.com` **is** loaded as a third party on more than three other domains. That is the whole purpose of it - to host JavaScript libraries and shared assets for other sites.

Also: `assets.example.com` is **not** blocked when visiting `www.example.com`.

**Comment by maranomynet** [comment]
Point is: This is an unwanted, and unexpected side-effect of using Google Analytics on a single website.
Damage is done, and the cookies are already out there. What can be done?

* Can Privacy Badger do something to prevent this from happening?
* Can admins of asset servers already hit by this do something to fix this on their end (and revert the block)?

**Comment by michael-oneill** [comment]
Well, the analytics cookies will be there (persistent, third-party, linked to targeted advertising .etc). If many PB users visit example.com then their PB will see assets.example.com as a tracking domain, which it is. The owner of example.com may not be tracking people but Google certainly is.
The problem is GA placing the cookies into the wildcard domain on example.com. I haven't checked recently but I thought there was a parameter to limit GA cookie placement to a specific domain, surely that's the way to fix it?

**Comment by maranomynet** [comment]
Google isn't tracking anything either - because BP has already blocked them. :-)

The default Google Analytics tracking code-example silently defaults to wildcard-domain cookies.

By reading boring API docs one learns about the "none" keyword which limits the cookie to the current domain only. (i.e. `ga('create', UA-xxxxxxx-x`, 'none');` or using a `cookieDomain:'none'` option)

This still leaves the old wildcard-domain cookies which now need to be removed manually...

```js
'_ga,_gat,_utma,__utmb,__utmt,__utmz'.split(',').forEach(function(c){
  document.cookie = c+'=;path=/;domain=.eplica.is;expires=Thu, 01 Jan 1970 00:00:00 GMT';
});
````

Which still doesn't help those unsuspecting users for which Privacy Badger has already decided poor old `assets.example.com` is an Evil Tracker™.

It would be nice if PB could change its heuristics as to detect/avoid cases like this - to help webmaster's avoid being DOS-ed by a rouge Google Analytics setup on a separate website.

**Comment by maranomynet** [comment]
Half-baked idea:
Problem would be solved if `assets.example.com` was only yellow-listed (cookie restricted).

**Comment by cooperq** [comment]
Right, but that only solves the problem for assets.example.com and we need to solve this in a more general way I think.

**Comment by maranomynet** [comment]
Is yellow-listing not an automatic/heuristic process? Is it just a manually maintained list?

**Comment by cooperq** [comment]
Yes, it is an actual manually maintained list.

**Comment by ghostwords** [comment]
Hi @maranomynet, could you provide an example (or two) of this situation in the wild?

**Comment by maranomynet** [comment]
The real-world example I was dealing with at the time, no longer exists because we ended up buying a new "assets" (cdn) domain and all links have been updated. (Not only because of this bug, since use of Privacy Badger isn't so widespread...)

But I believe the problem must still exist and shouldn't be difficult to replicate.

**Comment by ghostwords** [comment]
Yeah, CDN domains should avoid having any overlap with domains that set cookies, for performance, privacy and security reasons. Once the overlap happens, Privacy Badger users do have a problem. I'll keep an eye out for this coming up in practice. Not yet sure if there is a systematic solution, besides providing better tools for identifying why domains get blocked (#963) and an easier way to unblock mistakenly blocked domains for Privacy Badger users (#1645).

**Comment by ghostwords** [comment]
Google Analytics should now be reliably blocked as of Privacy Badger version 2019.7.1 (#367). This means this entire class of problem should be eliminated going forward (as GA will no longer be allowed to set the first party cookies in the first place). This can still be a problem with other, non-GA "first-party cookies/third-party assets that send those cookies" setups but I'm not aware of this being an issue outside of GA at this time.

**Comment by ghostwords** [comment]
Sort of related to #882.
